vms/avm/network/gossip.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/network/p2p/gossip"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/avm/txs"
+	"github.com/ava-labs/avalanchego/vms/avm/txs/mempool"
+)
+
+var (
+	_ p2p.Handler                  = (*txGossipHandler)(nil)
+	_ gossip.Set[*gossipTx]        = (*gossipMempool)(nil)
+	_ gossip.Gossipable            = (*gossipTx)(nil)
+	_ gossip.Marshaller[*gossipTx] = (*gossipTxParser)(nil)
+)
+
+// txGossipHandler is the handler called when serving gossip messages
+type txGossipHandler struct {
+	p2p.NoOpHandler
+	appGossipHandler  p2p.Handler
+	appRequestHandler p2p.Handler
+}
+
+func (t txGossipHandler) AppGossip(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	gossipBytes []byte,
+) {
+	t.appGossipHandler.AppGossip(ctx, nodeID, gossipBytes)
+}
+
+func (t txGossipHandler) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	deadline time.Time,
+	requestBytes []byte,
+) ([]byte, error) {
+	return t.appRequestHandler.AppRequest(ctx, nodeID, deadline, requestBytes)
+}
+
+type gossipTx struct {
+	tx *txs.Tx
```

Comment:
nit: we could probably add "origin" to this eventually for the future work on re-gossiping txs submitted over RPC

Reply:
not sure this is the best place to track if a tx was submitted over RPC or p2p. The mempool doesn't actually contain these `gossipTx` objects

Reply:
I removed this struct entirely (it really wasn't needed)